processor/k8sattributesprocessor/internal/kube/client.go

```diff
@@ -1761,69 +1749,161 @@ func needContainerAttributes(rules ExtractionRules) bool {
+		return
 	}
+	if rsView.UID == "" {
+		c.logger.Warn("add RS missing UID", zap.String("name", rsView.Name), zap.String("ns", rsView.Namespace))
+		return
+	}
+
+	// Centralized fallback with cache reuse
+	c.resolveReplicaSetDeploymentLinkage(&rsView)
+
+	c.upsertReplicaSet(rsView)
 }
 
-func (c *WatchClient) handleReplicaSetUpdate(_, newRS any) {
+func (c *WatchClient) handleReplicaSetUpdate(_, newObj any) {
 	c.telemetryBuilder.OtelsvcK8sReplicasetUpdated.Add(context.Background(), 1)
-	if replicaset, ok := newRS.(*apps_v1.ReplicaSet); ok {
-		c.addOrUpdateReplicaSet(replicaset)
-	} else {
-		c.logger.Error("object received was not of type apps_v1.ReplicaSet", zap.Any("received", newRS))
+
+	rsView, ok := normalizeRS(newObj)
+	if !ok || rsView.UID == "" {
+		c.logger.Warn("update received invalid RS", zap.Any("obj", newObj))
+		return
 	}
+
+	// Centralized fallback with cache reuse
+	c.resolveReplicaSetDeploymentLinkage(&rsView)
+
+	c.upsertReplicaSet(rsView)
 }
 
-func (c *WatchClient) handleReplicaSetDelete(obj any) {
-	c.telemetryBuilder.OtelsvcK8sReplicasetDeleted.Add(context.Background(), 1)
-	if replicaset, ok := ignoreDeletedFinalStateUnknown(obj).(*apps_v1.ReplicaSet); ok {
-		c.m.Lock()
-		key := string(replicaset.UID)
-		delete(c.ReplicaSets, key)
-		c.m.Unlock()
-	} else {
-		c.logger.Error("object received was not of type apps_v1.ReplicaSet", zap.Any("received", obj))
+// resolveReplicaSetDeploymentLinkage ensures rsView has DeploymentName/UID set
+// only when required by rules and when a controller Deployment owner can be confirmed.
+// It reuses cached linkage if present and performs at most one typed GET per RS UID.
+// If no controller Deployment is found, rsView.Deployment* remain empty.
+func (c *WatchClient) resolveReplicaSetDeploymentLinkage(rsView *replicasetView) {
+	// Only needed if deployment attrs are requested (and not name-from-RS)
+	needsDeployment := c.Rules.DeploymentUID ||
+		(c.Rules.DeploymentName && !c.Rules.DeploymentNameFromReplicaSet)
+	if !needsDeployment {
+		return
+	}
+	if rsView.DeploymentUID != "" {
+		return
 	}
-}
 
-func (c *WatchClient) addOrUpdateReplicaSet(replicaset *apps_v1.ReplicaSet) {
-	newReplicaSet := &ReplicaSet{
-		Name:      replicaset.Name,
-		Namespace: replicaset.Namespace,
-		UID:       string(replicaset.UID),
+	// Reuse cached linkage to avoid repeated GETs
+	c.m.RLock()
+	cached := c.ReplicaSets[rsView.UID]
+	c.m.RUnlock()
+	if cached != nil && cached.Deployment.UID != "" {
+		rsView.DeploymentName = cached.Deployment.Name
+		rsView.DeploymentUID = cached.Deployment.UID
+		return
 	}
 
-	for _, ownerReference := range replicaset.OwnerReferences {
-		if ownerReference.Kind == "Deployment" && ownerReference.Controller != nil && *ownerReference.Controller {
-			newReplicaSet.Deployment = Deployment{
-				Name: ownerReference.Name,
-				UID:  string(ownerReference.UID),
-			}
+	// Fallback: typed GET, only accept controller=true Deployment owner
```

Comment:
We prefer PartialObjectMetadata, but its ownerReferences can be missing or delayed. When deployment attributes are required, we first check our cache for previously resolved RS and if it is empty as last resort perform a one-time typed GET.